### ISSUE TITLE
Proxy doesn't work with kafka-clients:3.9.0

### ISSUE THREAD

**Issue by Pavelsky89** [kroxylicious/kroxylicious#1611]
**Describe the bug**
When using java kafka-clients (https://mvnrepository.com/artifact/org.apache.kafka/kafka-clients/3.9.0) producer in version 3.9.0 proxy is unable to handle requests.
We're getting following errors

> 	Caused by: org.apache.kafka.common.errors.UnsupportedVersionException: Can't size version 4 of ApiVersion
2024-11-20 08:54:22 <epollEventLoopGroup-5-1> WARN  io.kroxylicious.proxy.internal.KafkaProxyFrontendHandler:485 - Netty caught exception from the frontend: org.apache.kafka.common.errors.UnsupportedVersionException: Can't size version 4 of ApiVersion
io.netty.handler.codec.EncoderException: org.apache.kafka.common.errors.UnsupportedVersionException: Can't size version 4 of ApiVersion
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:125) ~[netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:893) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:875) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:984) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:868) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:115) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.kroxylicious.proxy.internal.ResponseOrderer.write(ResponseOrderer.java:69) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:891) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:875) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:984) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:868) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:964) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannel.write(AbstractChannel.java:300) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.kroxylicious.proxy.internal.KafkaProxyBackendHandler.channelRead(KafkaProxyBackendHandler.java:62) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.kroxylicious.proxy.internal.FilterHandler.forwardResponse(FilterHandler.java:370) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.handleResponseFilterResult(FilterHandler.java:243) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.lambda$configureResponseFilterChain$8(FilterHandler.java:194) ~[kroxylicious-runtime-0.9.0.jar:?]
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168) ~[?:?]
	at io.kroxylicious.proxy.internal.FilterHandler.configureResponseFilterChain(FilterHandler.java:194) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.readDecodedResponse(FilterHandler.java:176) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.channelRead(FilterHandler.java:105) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.kroxylicious.proxy.internal.FilterHandler.forwardResponse(FilterHandler.java:370) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.handleResponseFilterResult(FilterHandler.java:243) ~[kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.lambda$configureResponseFilterChain$8(FilterHandler.java:194) ~[kroxylicious-runtime-0.9.0.jar:?]
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:684) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:662) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2168) [?:?]
	at io.kroxylicious.proxy.internal.FilterHandler.configureResponseFilterChain(FilterHandler.java:194) [kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.readDecodedResponse(FilterHandler.java:176) [kroxylicious-runtime-0.9.0.jar:?]
	at io.kroxylicious.proxy.internal.FilterHandler.channelRead(FilterHandler.java:105) [kroxylicious-runtime-0.9.0.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) [netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) [netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1495) [netty-handler-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1358) [netty-handler-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1407) [netty-handler-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530) [netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469) [netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290) [netty-codec-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868) [netty-transport-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799) [netty-transport-classes-epoll-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501) [netty-transport-classes-epoll-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399) [netty-transport-classes-epoll-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) [netty-common-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.113.Final.jar:4.1.113.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.113.Final.jar:4.1.113.Final]
	at java.base/java.lang.Thread.run(Thread.java:840) [?:?]

**To Reproduce**
1. use proxy version 0.9.0
2. produce records using kafka-clients:3.9.0

**Expected behavior**
Records are passed to target kafka cluster

**Additional info**
* we're using TLS and SNI routing
*  target kafka cluster is 3.7.0
* connection from kafka-clients:3.9.0 to kafka cluster 3.7.0 without the proxy in between worked

**Comment by k-wall** [comment]
Thanks for reporting the issue.  We'll investigate soon.  #1550 should address the immediate issue but we'll want to understand why this is a breaking change.

**Comment by showuon** [comment]
I just tried to start up a pure kafka broker v3.8.1 without kroxy in between, and the producer and consumer with v3.9.0 can successfully write/read from the v3.8.1 brokers. So, Kafka should have done the api version negotiation correctly.

**Comment by k-wall** [comment]
> > https://issues.apache.org/jira/browse/KAFKA-17011 (3.9.0) introduced the new version of the ApiVersion (v4).
> 
> does that mean 3.9.0 is simply not supported by 0.9.0 proxy ?

Sorry you replied to a half written comment.

**Comment by k-wall** [comment]
> > I think it would be preferable if the proxy clamped the version number of the request to the highest it supports.
> 
> I thought we already did, I'm sure we've talked about this problem multiple times... So yes we should cap the API version at what the proxy supports.

The ApiVersion RPC is the mechanism that performs for the negotiation so client/broker exchange the other RPCs at mutually agreed version.   However, in this case we are talking about the ApiVersion RPC itself.
  

> 
> > If we do that the broker will see a v3 request and return a v3 response, which will reserialise okay. The Apache Kafka client seems to permit the arrival of the v3 response in response to its v4 request okay, however this approach would technically break the spec
> 
> Why does that break the spec? Isn't that the only way a 3.8.1 broker can respond to a v4 request?

If I've understood the broker's behaviour right, it uses the apiVersion of the ApiVersion request to determine the apiVersion of the ApiVersion response.   Send a v4 ApiVersion request to 3.8.0 Broker, you get a response claiming to be v4, even though 3.8.0 knows nothing about v4.  ApiVersion is a bit special!

**Comment by k-wall** [comment]
Note to self:  need to understand why the pre-3.9 Brokers don't fail when constructing the ApiVersions v4 response after the receipt of a v4 request.  The Broker must be using a different code path to construct this response than the Proxy does.  Perhaps the best fix is to make sure the Proxy mimics what the Broker does.

**Comment by k-wall** [comment]
Here's the difference, the Kafka Broker does this whilst serialising the unknown ApiVerions response.

https://github.com/apache/kafka/blob/23b85417b3ed53bf092877baef42608f7b778843/clients/src/main/java/org/apache/kafka/common/requests/RequestContext.java#L132

```java
    public short apiVersion() {
        // Use v0 when serializing an unhandled ApiVersion response
        if (isUnsupportedApiVersionsRequest())
            return 0;
        return header.apiVersion();
    }
```

**Comment by k-wall** [comment]
> > This gives us the possibility that the client may see two downgrades (one from the proxy then a second one from the broker). It looks like the Java client would tolerate this. I haven't checked the other clients.
> 
> 
> 
> Maybe we could do it in one downgrade. 

Yes agreed, I'd been thinking about it this evening and came to the same conclusion conclusion.

> If the proxy is aware of v3, and client and broker support v4.
> 
> 
> 
> 1. client sends proxy v4
> 
> 2. proxy detects it can't handle the version and forwards it's own v0 ApiVersions request to the broker


Rather than v0, it should use the maximum version the proxy can support.

> 
> 3. proxy intercepts v0 response, intersects in it's own supported versions and sets ErrorCode=UNSUPPORTED_VERSION

Agreed.  It might have the effect of further reducing the version range.


> 
> 4. client picks the new version and starts again (at this point it knows the highest ApiVersions version supported by proxy and broker)
> 
> 
> 
> edit: have put up a [PoC](https://github.com/kroxylicious/kroxylicious/pull/1619), a bit awkward to implement. 
> 
> 
> 
> I also tried doing it purely in a filter using `sendRequest` since that is able to control the outgoing api version (I don't think we can change apiVersion for a typical forwarded message). But something blows up in the statemachine when we try to shortcircuit respond `Can't size version 4 of ApiVersion. Increase log level to DEBUG for stacktrace`. I assume it's remembered the apiVersion of the request and using that version for the short circuit response.

I understand that issue.  It is because the proxy uses the request api version to serialise the response.   My PR should fix that.


> 
> 
> 
> The message decoder is unable to do the correlation since it's only handling decode, so there's a dirty hack to instruct another Filter to handle the correlation and error setting.

**Comment by robobario** [comment]
I had a play with another evil idea which was to forward an Opaque frame containing a hand-rolled ApiVersions request with the version set to short max. Idea being we pretend it's a client using a future version that should never come to be.

But from my read of the code this would be fragile as the Kafka code is inferring the Header version from the APIKey+version. In ApiMessageType it does:

``` java
public short requestHeaderVersion(short _version) {
        switch (apiKey) {
            ...
            case 18: // ApiVersions
                if (_version >= 3) {
                    return (short) 2;
                } else {
                    return (short) 1;
                }
```

so in future if/when further Header Versions were added, faking a large version could mean the broker thinks the message is using the latest header version but the proxy hasn't accounted for that.

TL:DR; pretending the proxy is ahead of the broker to prompt the error reaction is a bad idea. It was just tempting because we don't need to correlate anything in the Proxy.

**Comment by robobario** [comment]
>Rather than v0, it should use the maximum version the proxy can support.

why forward on the highest proxy supported version? I guess we can pass more information to the broker that way, like the clientId in the header and client software fields in the value.  Better for debugging.

We still have to correlate the response and ensure we return the error code to let the client know it was using an unsupported version, the error might be set by the broker or the proxy depending if the proxy is ahead of the broker..

**Comment by k-wall** [comment]
> > Rather than v0, it should use the maximum version the proxy can support.
> 
> why forward on the highest proxy supported version? I guess we can pass more information to the broker that way, like the clientId in the header and client software fields in the value. Better for debugging.

My rationale if that sending the highest is least surprise.  It is what the broker expects its peer to do.

> 
> We still have to correlate the response and ensure we return the error code to let the client know it was using an unsupported version, the error might be set by the broker or the proxy depending if the proxy is ahead of the broker.

Agree on both those points.

**Comment by k-wall** [comment]
@robobario I didn't get chance to continue with this today, I'll follow up on Monday.

**Comment by showuon** [comment]
> So my read of it is that the broker cannot decode arbitrary future ApiVersionsData messages, if a new RequestHeaderData version is added with an extra field, the broker would not be aware how to skip over the new header field. So instead it reads the part of the header it is sure will be present and avoids decoding the body.

Rob is correct.

About this issue, I think we _should_ support apiVersion v4 to allow the apiVersionResponse contains supported version 0 to allow the clients to successfully deserialize it. This is a bug fix and we should include it. Of course we can think about how we can support "future" api version request and still succesfully work in kroxy. 
More context: https://lists.apache.org/thread/7wx3j788ztf1vl7xm9ryt8cwl3tk9lw3

That is, in v3.9.0 client, we should be able to see this without error: (min version: 0, finalized version: 0)
```
> bin/kafka-features.sh --bootstrap-controller localhost:9093 describe

Feature: kraft.version  SupportedMinVersion: 0  SupportedMaxVersion: 1  FinalizedVersionLevel: 0 Epoch: 5
```

doc: https://kafka.apache.org/documentation/#kraft_reconfig

**Comment by robobario** [comment]
Have had a quick check and the proxy (we support Kafka 3.9 in main) looks to do the right thing if I configure it to point at a 3.9 controller:

```
bin/kafka-features.sh --bootstrap-controller localhost:9192 describe
Feature: kraft.version	SupportedMinVersion: 0	SupportedMaxVersion: 1	FinalizedVersionLevel: 0	Epoch: 10
Feature: metadata.version	SupportedMinVersion: 3.0-IV1	SupportedMaxVersion: 3.9-IV0	FinalizedVersionLevel: 3.9-IV0	Epoch: 10
```
vs direct to the controller
```
robeyoun@robeyoun-thinkpadp16vgen1:~/Downloads/kafka_2.13-3.9.0$ bin/kafka-features.sh --bootstrap-controller localhost:9093 describe
Feature: kraft.version	SupportedMinVersion: 0	SupportedMaxVersion: 1	FinalizedVersionLevel: 0	Epoch: 10
Feature: metadata.version	SupportedMinVersion: 3.0-IV1	SupportedMaxVersion: 3.9-IV0	FinalizedVersionLevel: 3.9-IV0	Epoch: 10
```

good thing to raise though, should we be integration testing proxy in front of controllers too?

**Comment by k-wall** [comment]
> Another thing: the Kafka broker doesn't attempt to decode requests that are ahead, see [RequestContext](https://github.com/apache/kafka/blob/3f834781a4311fb26fec740ee68bddb70a6f0701/clients/src/main/java/org/apache/kafka/common/requests/RequestContext.java#L113).


> Another thing: the Kafka broker doesn't attempt to decode requests that are ahead, see [RequestContext](https://github.com/apache/kafka/blob/3f834781a4311fb26fec740ee68bddb70a6f0701/clients/src/main/java/org/apache/kafka/common/requests/RequestContext.java#L113).

^^ I'd completely missed that fact.

**Comment by k-wall** [comment]
Reflecting on this, I think we should bring this issue into a discussion so we end up with a clear statement which we implement from.   I think there's a fair of subtlety that could be otherwise lost.  Let's talk through how each case needs to be handled (and talk through any options).

**Comment by k-wall** [comment]
#1635 should address this issue.
